File: latte/src/main/java/gg/beemo/latte/broker/rabbitmq/RabbitConnection.kt
package gg.beemo.latte.broker.rabbitmq

import com.rabbitmq.client.*
import gg.beemo.latte.broker.BrokerConnection
import gg.beemo.latte.broker.BrokerMessageHeaders
import gg.beemo.latte.broker.MessageId
import gg.beemo.latte.logging.Log
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.util.*
import java.util.concurrent.atomic.AtomicBoolean
import javax.net.ssl.SSLContext


private class ChannelData(
    val channel: Channel,
    val sendMutex: Mutex = Mutex(),
    var isConsuming: AtomicBoolean = AtomicBoolean(false),
    var consumerTag: String? = null,
)

class RabbitConnection(
    rabbitHosts: Array<String>,
    override val serviceName: String,
    override val instanceId: String,
    private val useTls: Boolean = false,
    private val username: String = "guest",
    private val password: String = "guest",
) : BrokerConnection() {

    override val supportsTopicHotSwap = true
    override val deferInitialTopicCreation = true
    private val rabbitAddresses = rabbitHosts.map(Address::parseAddress)
    private val log by Log

    private var connection: Connection? = null
    private val channels = Collections.synchronizedMap(HashMap<String, ChannelData>())

    override suspend fun abstractStart() {
        connection = ConnectionFactory().also {
            if (useTls) {
                it.useSslProtocol(SSLContext.getDefault())
                it.enableHostnameVerification()
            }
            it.useNio()
            it.username = username
            it.password = password
            it.isAutomaticRecoveryEnabled = true
            it.isTopologyRecoveryEnabled = true
        }.newConnection(rabbitAddresses, instanceId)
    }

    override suspend fun abstractSend(
        topic: String,
        key: String,
        value: String,
        headers: BrokerMessageHeaders
    ): MessageId {
        if (shouldDispatchExternallyAfterShortCircuit(topic, key, value, headers)) {

            val channelData = getOrCreateChannel(topic)
            // RabbitMQ's channels are not thread-safe for sending. Consuming and sending
            // through the same channel at the same time is fine though.
            channelData.sendMutex.withLock {
                val properties = AMQP.BasicProperties.Builder().apply {
                    // https://www.rabbitmq.com/docs/publishers#message-properties
                    deliveryMode(2) // Persistent
                    headers(headers.headers) // lol
                }.build()
                channelData.channel.basicPublish(topic, key, properties, value.toByteArray())
            }

        }

        return headers.messageId
    }

    override fun destroy() {
        log.debug("Destroying RabbitConnection")
        connection?.close()
        connection = null
        super.destroy()
    }

    override fun createTopic(topic: String) {
        val channelData = getOrCreateChannel(topic)
        if (channelData.isConsuming.getAndSet(true)) {
            return
        }
        val consumer = object : DefaultConsumer(channelData.channel) {

            override fun handleDelivery(
                consumerTag: String,
                envelope: Envelope,
                properties: AMQP.BasicProperties,
                body: ByteArray
            ) {
                val key = envelope.routingKey ?: ""
                try {
                    val value = String(body)
                    val headers = BrokerMessageHeaders(properties.headers.mapValues { it.value.toString() })
                    dispatchIncomingMessage(topic, key, value, headers)
                } catch (e: Exception) {
                    log.error("Error handling incoming broker message in topic $topic, dropping message", e)
                }
                channel.basicAck(envelope.deliveryTag, false)
            }

            override fun handleShutdownSignal(consumerTag: String, sig: ShutdownSignalException) {
                if (sig.isInitiatedByApplication) {
                    return
                }
                log.error("RabbitMQ consumer for topic $topic has shut down unexpectedly", sig)
                // The client _should_ automatically recover the connection
            }
        }
        channelData.consumerTag = channelData.channel.basicConsume(createQueueName(topic), false, consumer)
    }

    override fun removeTopic(topic: String) {
        val channel = channels.remove(topic)
        channel?.channel?.queueDelete(createQueueName(topic))
        channel?.channel?.close()
    }

    private fun getOrCreateChannel(topic: String): ChannelData {
        return channels.computeIfAbsent(topic) {
            val connection = checkNotNull(connection) { "Connection not open" }
            val channel = connection.createChannel().apply {
                val exchangeName = topic
                val queueName = createQueueName(topic)
                val routingKey = "#"
                exchangeDeclare(exchangeName, BuiltinExchangeType.TOPIC, true)
                queueDeclare(queueName, true, false, false, null)
                queueBind(queueName, exchangeName, routingKey)
            }
            ChannelData(channel)
        }
    }

    private fun createQueueName(topic: String) = "$serviceName.$instanceId.$topic"

}

File: water/src/broker/rabbitmq/RabbitConnection.ts
import type { Consumer, Publisher } from "rabbitmq-client";
import { Connection } from "rabbitmq-client";
import type { MessageId } from "../BrokerConnection.js";
import { BrokerConnection } from "../BrokerConnection.js";
import { BrokerMessageHeaders } from "../BrokerMessageHeaders.js";
import { Logger } from "../../logging/Logger.js";

export class RabbitConnection extends BrokerConnection {

    private static readonly TAG = "RabbitConnection";

    public override supportsTopicHotSwap: boolean = true;
    public override deferInitialTopicCreation: boolean = true;

    private connection: Connection | null = null;
    private readonly consumers: Map<string, Consumer> = new Map();
    private readonly publishers: Map<string, Publisher> = new Map();

    public constructor(
        private readonly rabbitHosts: string[],
        public override readonly serviceName: string,
        public override readonly instanceId: string,
        private readonly useTls: boolean = false,
        private readonly username: string = "guest",
        private readonly password: string = "guest",
    ) {
        super();
    }

    public override async abstractStart(): Promise<void> {
        this.connection = new Connection({
            hosts: this.rabbitHosts,
            username: this.username,
            password: this.password,
            tls: this.useTls,
            connectionName: `${this.serviceName}: ${this.instanceId}`,
        });
        this.connection.on("connection.blocked", reason => {
            Logger.warn(RabbitConnection.TAG, `RabbitMQ server blocked connection for reason: ${reason}`);
        });
        this.connection.on("connection.unblocked", () => {
            Logger.info(RabbitConnection.TAG, "RabbitMQ server unblocked connection");
        });
        await new Promise<void>((resolve, reject) => {
            // Use `on` instead of `once` to re-use the same logging code for future connection updates
            this.connection!.on("connection", () => {
                Logger.info(RabbitConnection.TAG, "RabbitMQ connection (re)established");
                resolve();
            });
            this.connection!.on("error", error => {
                Logger.error(RabbitConnection.TAG, "Error in RabbitMQ connection", error);
                reject(error);
            });
        });
    }

    public override async destroy(): Promise<void> {
        await super.destroy();
        await this.connection?.close();
    }

    public override async abstractSend(topic: string, key: string, value: string, headers: BrokerMessageHeaders): Promise<MessageId> {
        if (this.shouldDispatchExternallyAfterShortCircuit(topic, key, value, headers)) {
            let publisher = this.publishers.get(topic);
            if (!publisher) {
                publisher = this
                    .ensureConnection()
                    .createPublisher(this.commonPublisherConsumerArguments(topic));
                this.publishers.set(topic, publisher);
            }
            await publisher.send({
                exchange: topic,
                routingKey: key,
                headers: headers.headers,
                messageId: headers.messageId,
                durable: true,
            }, value);
        }
        return headers.messageId;
    }

    protected override createTopic(topic: string): void {
        if (this.consumers.has(topic)) {
            return;
        }
        const queueName = this.createQueueName(topic);
        this.consumers.set(topic, this.ensureConnection().createConsumer({
            queue: queueName,
            ...this.commonPublisherConsumerArguments(topic),
        }, async (msg, reply) => {
            const key = msg.routingKey;
            const value = msg.body instanceof Buffer ? msg.body.toString("utf8") : String(msg.body);
            const headers = new BrokerMessageHeaders(msg.headers ?? {});
            this.dispatchIncomingMessage(topic, key, value, headers);
        }));
    }

    protected override removeTopic(topic: string): void {
        const consumer = this.consumers.get(topic);
        if (consumer) {
            consumer.close().catch(e => {
                Logger.error(RabbitConnection.TAG, `Error closing consumer for topic ${topic}`, e);
            });
            this.consumers.delete(topic);
        }
    }

    private ensureConnection(): Connection {
        const connection = this.connection;
        if (!connection) {
            throw new Error("Connection not open");
        }
        return connection;
    }

    private createQueueName(topic: string): string {
        return `${this.serviceName}.${this.instanceId}.${topic}`;
    }

    private commonPublisherConsumerArguments(topic: string) {
        const exchangeName = topic;
        const queueName = this.createQueueName(topic);
        const routingKey = "#";
        return {
            queues: [{
                queue: queueName,
                durable: true,
                exclusive: false,
                autoDelete: false,
            }],
            exchanges: [{
                exchange: exchangeName,
                type: "topic",
                durable: true,
            }],
            queueBindings: [{
                queue: queueName,
                exchange: exchangeName,
                routingKey,
            }],
        };
    }

}
